src/main/java/org/example/jackson/bench/NumberUtil.java

```diff
@@ -66,4 +68,108 @@ public static long alternateParseLong(String s) {
         return neg ? -num : num;
     }
 
+    public static long alternateParseSimplified(String s) {
+        /* Ok: let's keep strategy simple: ignoring optional minus sign,
+         * we'll accept 1 - 19 digits and parse things efficiently;
+         * otherwise just defer to JDK parse functionality.
+         */
+        int len = s.length();
+        boolean neg = s.charAt(0) == '-';
+        int offset;
+        // must have 1 - 19 digits after optional sign:
+        // negative?
+        if (neg) {
+            if (len == 1 || len > 20) {
+                return Long.parseLong(s);
+            }
+            offset = 1;
+        } else {
+            if (len > 19) {
+                return Long.parseLong(s);
+            }
+            offset = 0;
+        }
+        long num = 0L;
+        for (int i = offset; i < len; i++) {
+            char c = s.charAt(i);
+            if (c > '9' || c < '0') {
+                return Long.parseLong(s);
+            }
+            num = (num * 10) + (c - '0');
```

Comment:
when i == len, could we use Math.addExact to fail if there is an overflow or a variant like this?

```
    private static long addExact(final String s, final long x, final long y) {
        long r = x + y;
        // HD 2-12 Overflow iff both arguments have the opposite sign of the result
        if (((x ^ r) & (y ^ r)) < 0) {
            throw new NumberFormatException("For input string: " + s);
        }
        return r;
    }
```

we should really only reach overflow point when handling the final digit

or since we know that x and y are positive based on how the calling code works, we could just check if x+y<0 to see if there is an overflow.

```
    private static long addExact(final String s, final long x, final long y) {
        long r = x + y;
        if (r < 0) {
            throw new NumberFormatException("For input string: " + s);
        }
        return r;
    }
```

Reply:
I have added this change to main branch to demo it

Reply:
Yeah, this shouldn't be understood as a proposal for upstream in Jackson. It's intended as a way to explore our options.

What I'm struggling most with is what the exact contract of `#parseLong(String)` should be. How much validation checking the method has to do versus how much validation the caller has to do or has already done. E.g. do we really have to check if every digit is numeric or has that already been done? Same for overflow. The `char[]` based variant does not have that many callers and was easy to analyze. The `String` based variant has quite a bit more callers and is harder to analyze.

One this to keep in mind is that `#addExact` is an intrinsic. It likely will be replaced with assembly code using `JO`/`JNO`.

Reply:
unfortunately, addExact throws an ArithmeticException on overflow - while for backward compatibility parseLong should throw a NumberFormatException - so we would need to catch ArithmeticException and throw NumberFormatException

parseLong right now should concentrate on simple numbers with 0-9 and +/- --> we're happy to have reduced performance for hexadecimal, etc.

Reply:
@marschall do you think my small mod to your alternateJacksonParseSimplified is a good way to go. This version performs well in beanchmarks

The switch version is nice but I'm not sure there is an easy way to get it to fail over to Long.parseLong for non-digit chars.